process/missingData/missingDataResolver.go

```diff
@@ -231,6 +233,102 @@ func (r *Resolver) WaitForMissingData(timeout time.Duration) error {
 	}
 }
 
+// RequestMissingShardHeadersBlocking requests the missing shard headers and proofs for the given meta header and waits for results.
+func (r *Resolver) RequestMissingShardHeadersBlocking(
+	metaHeader data.MetaHeaderHandler,
+	timeout time.Duration,
+) error {
+	err := r.RequestMissingShardHeaders(metaHeader)
+	if err != nil {
+		return err
+	}
+
+	return r.WaitForMissingData(timeout)
+}
+
+// RequestMissingShardHeaders requests the missing shard headers and proofs for the given meta header.
+func (r *Resolver) RequestMissingShardHeaders(
+	metaHeader data.MetaHeaderHandler,
+) error {
+	if check.IfNil(metaHeader) {
+		return process.ErrNilMetaBlockHeader
+	}
+
+	if metaHeader.IsStartOfEpochBlock() && metaHeader.GetEpochStartHandler() != nil {
+		r.requestEpochStartLastFinalizedHeaders(metaHeader.GetEpochStartHandler())
+	}
+
+	shardDataProposedNonces := make(map[uint32]uint64)
+	for _, shardProposalData := range metaHeader.GetShardInfoProposalHandlers() {
+		shardID := shardProposalData.GetShardID()
+		shardDataProposedNonces[shardID] = shardProposalData.GetNonce()
+
+		r.requestHeaderIfNeeded(shardID, shardProposalData.GetHeaderHash())
+		r.requestProofIfNeeded(shardID, shardProposalData.GetHeaderHash())
+	}
+
+	shardDataFinalizedNonces := getShardDataFinalizedNonces(metaHeader.GetShardInfoHandlers())
+	r.requestNonceGapsIfNeeded(shardDataFinalizedNonces, shardDataProposedNonces)
+
+	return nil
+}
+
+func (r *Resolver) requestEpochStartLastFinalizedHeaders(epochStartHandler data.EpochStartHandler) {
+	for _, finalizedHdr := range epochStartHandler.GetLastFinalizedHeaderHandlers() {
+		r.requestHeaderIfNeeded(finalizedHdr.GetShardID(), finalizedHdr.GetHeaderHash())
+		r.requestProofIfNeeded(finalizedHdr.GetShardID(), finalizedHdr.GetHeaderHash())
+	}
+}
+
+func getShardDataFinalizedNonces(shardInfoHandlers []data.ShardDataHandler) map[uint32]uint64 {
+	shardDataFinalizedNonces := make(map[uint32]uint64)
+	for _, shardData := range shardInfoHandlers {
+		shardDataFinalizedNonces[shardData.GetShardID()] = shardData.GetNonce()
+	}
+	return shardDataFinalizedNonces
+}
+
+func (r *Resolver) requestNonceGapsIfNeeded(shardDataFinalizedNonces, shardDataProposedNonces map[uint32]uint64) {
+	for shardID, proposedNonce := range shardDataProposedNonces {
+		lastFinalizedNonce, found := shardDataFinalizedNonces[shardID]
+		if !found {
+			log.Warn("Resolver.requestNonceGapsIfNeeded: shard not found in shardDataFinalizedNonces", "shard", shardID)
```

Comment:
i do not think this warn is good. Like it could appear in a lof of cases. 
like if no new shardData is finalized, but you have new proposals.

Reply:
Right, I also thought it's not a good idea, but was not sure if this was also a valid case. Removed the warn